text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,236 @@
+Cargo passes in all dependencies it builds with `--extern`.
+However Cargo does not know about the standard library, so builds of it are taken from the sysroot.
+
+For cross-compiling, one can often download standard library binaries with rustup.
+This is convenient, but one cannot expect pre-built binaries for all platforms.
+In particular, embedded systems often have detailed configurations to convey as much information as possible about the hardware to the compiler.
+Furthermore, not all of the stdlib is available on every platform---there is an RFC in the works to pre-build a smaller set of crates for [Cortex-M microcontrollers](https://github.com/rust-lang/rfcs/pull/1645).
+It would be nice to know if the available subset is adequate before attempting a build.
+We can only do that if all packages have explicit standard library deps to cross-reference with platform requirements of each standard library crate.
+
+Now rustup could be augmented to build stdlib binaries in addition to downloading them, but we have extra configuration options for the standard library such as [panic strategies](https://github.com/rust-lang/rfcs/blob/master/text/1513-less-unwinding.md) along with plans to add more such as [`core` without floats](https://github.com/rust-lang/rfcs/issues/1364).
+One could also cobble together a way to to tell rustup such configuration options, but we'd like to make sure that all dependencies agree with the plan before trying to execute it.
+For panicking, packages can already set a `profile.dev.panic` option to require a specific strategy, and for float support we should add a (default opt-in) feature to core.
+Then, if packages can explicitly depend on core they can also specify whether the float feature is needed.
+Cargo would be able to infer both of these options by inspecting all crates in the dependency graph.
+
+Rustbuild must currently perform multiple `cargo builds`, the first to build the standard library and the rest to build things which depend on the standard library.
+If rustc, rustfmt, etc, and their deps (some of which come from crates.io, and thus aren't specially tailored for building rust) declare deps on std, rustbuild wouldn't need multiple lockfiles.
+Keeping multiple lockfiles in sync is a nuisance, and this gets us one step closer to a single `cargo build` building rustc.
+
+The use-cases so far mainly benefit niche corners of the Rust community, but the last should be useful for just about everyone.
+Now that multiple versions of Rust have been released, it can be useful to specify the minimum version.
+If and when Rust 2, a version with breaking changes, comes out, this will be all the more important.
+We don't yet have a plan yet to track the language itself, but by tracking standard library dependencies we make it trivial to specify version requirements like any other package.
```

Comment:
This last sentence is hard to understand. What does it mean to "track the language itself" in this context? Could be more plainly stated like "Because the versions of the crates in the sysroot are required to be the same as the version of the language supported by the compiler, specifying the semver requirements of a crate in the sysroot effectively specifies the semver requirements of the language itself".